File: poc-ddd/src/domain/@shared/event/event.dispatcher.impl.ts
import EventDispatcher from "./event.dispatcher";
import EventHandler from "./event.handler";
import Event from "./event";

export default class EventDispatcherImpl implements EventDispatcher {
    private eventHandlers = new Map<string, EventHandler[]>();

    notify(event: Event): void {
        // O evento é o nome da classe (event.constructor.name)
        const handlers = this.eventHandlers.get(event.constructor.name) 
        if (handlers) {
            handlers.forEach(eventHandler => eventHandler.handle(event))
        }
    }

    register(eventName: string, handler: EventHandler): void {
        const handlers = this.eventHandlers.get(eventName)
        if (handlers) {
            handlers.push(handler);
        } else {
            this.eventHandlers.set(eventName, [handler])
        }
    }

    unregister(eventName: string, handler: EventHandler): void {
        const handlers = this.eventHandlers.get(eventName)
        if (handlers) {
            const updatedHandlers = handlers.filter(h => h !== handler)
            this.eventHandlers.set(eventName, updatedHandlers);
        }
    }

    unregisterAll(): void {
        this.eventHandlers.clear();
    }

    getHandlers(handler: string): EventHandler[] {
        const handlers = this.eventHandlers.get(handler)
        if (handlers && handlers.length > 0) {
            return handlers;
        }
        return [];
    }

}

File: poc-microsservices/wallet-shared/src/main/java/aristides/dev/shared/events/EventDispatcherImpl.java
package aristides.dev.shared.events;

import aristides.dev.shared.exceptions.DuplicateHandlerException;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class EventDispatcherImpl implements EventDispatcher {

    private final Map<String, List<EventHandler>> eventHandlers = new HashMap<>();

    @Override
    public void register(String eventName, EventHandler handler) {
        List<EventHandler> handlers = eventHandlers.get(eventName);
        if (handlers == null) {
            handlers = new ArrayList<>();
        } else {
            for (EventHandler existingHandler : handlers) {
                if (existingHandler.getClass().equals(handler.getClass())) {
                    throw new DuplicateHandlerException(handler);
                }
            }
        }
        handlers.add(handler);
        eventHandlers.put(eventName, handlers);
    }

    @Override
    public void dispatch(BaseEvent event) {
        List<EventHandler> handlers = eventHandlers.get(event.name());
        if (handlers != null) {
            for (EventHandler eventHandler : handlers) {
                eventHandler.handle(event); //TODO PONTO DE MELHORIA. Executar em paralelo
            }
        }
    }

    @Override
    public void remove(String eventName, EventHandler handler) {
        List<EventHandler> handlers = eventHandlers.get(eventName);
        if (handlers != null) {
            handlers.removeIf(existingHandler -> existingHandler.getClass().equals(handler.getClass()));
        }
    }

    @Override
    public Boolean has(String eventName, EventHandler handler) {
        List<EventHandler> handlers = eventHandlers.get(eventName);
        if (handlers == null) {
            return false;
        }
        return handlers.stream().anyMatch(existingHandler -> existingHandler.getClass().equals(handler.getClass()));
    }

    @Override
    public void clear() {
        this.eventHandlers.clear();
    }
}
